### PR TITLE
fix #2269 Add test util to capture logs in early-established loggers

### LINKED ISSUE #2269: Explore a utility LoggerFactory to allow late binding of Loggers in tests
# Motivation
From time to time it can be useful to assert logging in tests. For example, to ensure that an operator encountering an extraneous `onError` goes through the default hook behavior - given that default behavior is to only log.

The trouble is that in such cases, the actual `Logger` is one that is statically initialized on a utility class, which is very likely to be classloaded earlier during the test run. So there's no guarantee that eg. `Loggers.useCustomLogger(ignored -> new TestLogger())` would work.

## Desired solution
A `reactor-test` way to capture log output of some tests without depending on any particular logging framework, and without forcing the whole test set to output to an in memory collection (ie logs should still go to logback for the most part).

A possible way of doing that would be to complement `TestLogger` in `reactor-test` with a `Loggers` factory that would by default forward to the current `Loggers` factory, but offer an on-demand way of diverting the logs to an assertable datastructure.

Ie. install the logger factory early (and once), but pinpoint where it capture logs in individual tests / `@Before`. Plus of course the possibility to `reset()` it to forget captured logs from one test to another.

## Considered alternatives
In #2205 the `LoggerUtils` was introduced but it relies on the fact that `logback` is configured and on the test classpath. It allows to set and unset a custom `Appender` that additively adds to a `TestLogger`. This solution cannot be generalized to `reactor-test`, which cannot enforce a dependency to slf4j nor logback.

### PR DESCRIPTION
**Pull request by ericbottard**

### CODE REVIEW

**Comment by ericbottard** [comment]
I subverted `LoggerUtils` for now to prove the POC.

Please take a look @simonbasle and if we're happy with the approach I can polish.

File: reactor-core/src/main/java/reactor/util/Loggers.java [hunk 1]
```diff
@@ -64,7 +64,7 @@
 	 * @see #useConsoleLoggers()
 	 * @see #useVerboseConsoleLoggers()
 	 */
-	public static final void resetLoggerFactory() {
+	public static void resetLoggerFactory() {
```
Comment (simonbasle) [inline reactor-core/src/main/java/reactor/util/Loggers.java:67]: why removing the `final`s in this file ?

File: reactor-core/src/test/java/reactor/core/publisher/OperatorsTest.java [hunk 2]
```diff
@@ -972,4 +972,11 @@ public void reportThrowInSubscribeWithFuseableErrorResumed() {
 
 		assertSubscriber.assertNoError().awaitAndAssertNextValues(123);
 	}
+
+	@Test
+	public void onDiscardCallbackErrorsLog() {
+		Context context = Operators.enableOnDiscard(Context.empty(), t -> {throw new RuntimeException("Boom");});
+		Operators.onDiscard("Foo", context);
+
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/core/publisher/OperatorsTest.java:980]: (need a TODO to complete this test, no assertion)

File: reactor-core/src/test/java/reactor/test/LoggerUtils.java [hunk 3]
```diff
@@ -1,59 +1,208 @@
...
 
 	public static void addAppender(TestLogger testLogger, Class<?> classWithLogger) {
-		org.slf4j.Logger slf4jLogger = LoggerFactory.getLogger(classWithLogger);
-		if (slf4jLogger instanceof Logger) {
-			Logger logbackLogger = (Logger) slf4jLogger;
-			TestAppender appender = new TestAppender(testLogger);
-			appender.start();
-			logbackLogger.addAppender(appender);
-		}
+		LoggerUtils.testLogger = testLogger;
```
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/test/LoggerUtils.java:19]: I think there is still value to _provide_ a `TestLogger` as opposed to letting the method create one internally and return it. see comment about tuning logging levels.

**Review by simonbasle — CHANGES_REQUESTED** [review]
(I've fixed the empty `ErrorFormatterTest.java` separately, please rebase this PR on the latest master)

File: reactor-core/src/test/java/reactor/test/LoggerUtils.java [hunk 4]
```diff
@@ -62,7 +68,7 @@ public String getName() {
 
 		@Override
 		public boolean isTraceEnabled() {
-			return delegate.isTraceEnabled();
+			return delegate.isTraceEnabled() || (LoggerUtils.testLogger != null && LoggerUtils.testLogger.isTraceEnabled());
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/test/LoggerUtils.java:71]: for the condition, capture the `testLogger` into a local variable (the `testLogger` could be nulled out between the `!= null` test and the `isTraceEnabled()` call)

File: reactor-core/src/test/java/reactor/test/LoggerUtils.java [hunk 5]
```diff
@@ -1,59 +1,214 @@
...
 		}
 
 		@Override
-		protected void append(ILoggingEvent eventObject) {
-			if (eventObject.getLevel() == Level.ERROR) {
-				testLogger.error(eventObject.getFormattedMessage()
-				                            .concat("\n")
-				                            .concat(throwableProxyConverter.convert(eventObject)));
+		public void trace(String msg) {
+			if (LoggerUtils.testLogger != null) {
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/test/LoggerUtils.java:76]: same comment as above about `testLogger` capture

File: reactor-core/src/test/java/reactor/test/LoggerUtils.java [hunk 6]
```diff
@@ -1,59 +1,214 @@
...
+	}
+
+	/**
+	 * Sets a {@link Loggers#useCustomLoggers(Function) logger factory} that will return loggers that not only use the
+	 * original logging framework used by reactor, but also use the logger set via {@link #installAdditionalLogger(Logger)}, irrespective
+	 * of its name or how it was obtained. The expectation here is that tests that want to assess that something is
+	 * logged by reactor will pass a {@link TestLogger} instance to {@link #installAdditionalLogger(Logger)}, trigger the operation
+	 * under scrutiny, assert the logger contents and reset state by calling {@link #resetAdditionalLogger()}.
+	 */
+	public static void setupDivertingLoggerFactory() {
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/test/LoggerUtils.java:30]: I'm on the fence about the method actually setting up the logger factory, vs returning a "factory" (as a `Function`) to pass to `Loggers.useCustomLoggers`.

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/test/java/reactor/ReactorTestExecutionListener.java [hunk 7]
```diff
@@ -24,10 +25,21 @@
...
+ *     <li>resets {@link Schedulers} related infrastructure, making sure no dirty state remains,</li>
+ *     <li>installs custom assertJ {@link Representation} for some of reactor types,</li>
+ *     <li>installs a custom {@link Logger} factory <strong>very</strong> early in the suite lifecycle, so that loggers
+ *     in reactor (which are typically static members initialized early) can be diverted and asserted in tests.</li>
+ * </ul>
+ */
 public class ReactorTestExecutionListener implements TestExecutionListener {
 
-	public static void reset() {
+	private static void resetHooksAndSchedulers() {
```
Comment (simonbasle) [inline reactor-core/src/test/java/reactor/ReactorTestExecutionListener.java:42]: still needs to remove the call to old `reset()` method in `BaseOperatorTest`

File: reactor-test/src/main/java/reactor/test/util/LoggerUtils.java [hunk 8]
```diff
@@ -0,0 +1,281 @@
...
+				}
+				lfField.set(Loggers.class, orginalFactory);
+			}
+			catch (NoSuchFieldException | IllegalAccessException e) {
+				throw new RuntimeException(e);
+			}
+		}
+	}
+
+	public static void enableCaptureWith(Logger testLogger) {
```
Comment (simonbasle) [inline reactor-test/src/main/java/reactor/test/util/LoggerUtils.java:89]: nitpick: I'd put the static methods above both static inner classes, and add basic documentation

File: reactor-test/src/test/java/reactor/test/util/LoggerUtilsTest.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+		}
+	}
+
+	@Test
+	void disposeOnlyUninstallsItelf() {
+		Disposable disposable = LoggerUtils.useCurrentLoggersWithCapture();
+		assertThatExceptionOfType(IllegalStateException.class).isThrownBy(() -> {
+			Loggers.resetLoggerFactory(); // Overwrites our custom logger
+			disposable.dispose();
+		});
```
Comment (simonbasle) [inline reactor-test/src/test/java/reactor/test/util/LoggerUtilsTest.java:58]: the message should be asserted

**Review by simonbasle — CHANGES_REQUESTED** [review]

File: reactor-core/src/test/java/reactor/test/LoggerUtils.java [hunk 10]
```diff
@@ -1,59 +1,208 @@
...
-		protected void append(ILoggingEvent eventObject) {
-			if (eventObject.getLevel() == Level.ERROR) {
-				testLogger.error(eventObject.getFormattedMessage()
-				                            .concat("\n")
-				                            .concat(throwableProxyConverter.convert(eventObject)));
+		public void trace(String msg) {
+			if (LoggerUtils.testLogger != null) {
+				LoggerUtils.testLogger.trace(msg);
+			}
+			delegate.trace(msg);
```
[the remark this answers is not in the document]
Comment (simonbasle) [reply reactor-core/src/test/java/reactor/test/LoggerUtils.java]: defering this until somebody actually "complains" about the double output

**Review by simonbasle — APPROVED** [review]

**Pull request merged**